src/validation.cpp

```diff
@@ -3650,8 +3651,8 @@ bool ChainstateManager::ProcessNewBlockHeaders(const std::vector<CBlockHeader>&
             }
         }
     }
-    if (NotifyHeaderTip()) {
-        if (::ChainstateActive().IsInitialBlockDownload() && ppindex && *ppindex) {
+    if (NotifyHeaderTip(ActiveChainstate())) {
```

Comment:
`ActiveChainstate()` doesn't take `cs_main`, whereas `::ChainstateActive()` does. Is that a problem? I imagine we shouldn't be reading `m_active_chainstate` without `cs_main`. The comment says:

```
    //! Once this pointer is set to a corresponding chainstate, it will not
    //! be reset until init.cpp:Shutdown(). This means it is safe to acquire
    //! the contents of this pointer with ::cs_main held, release the lock,
    //! and then use the reference without concern of it being deconstructed.
```

Should `m_active_chainstate` be GUARDED_BY cs_main and `ActiveChainstate()` have EXCLUSIVE_LOCKS_REQUIRED(::cs_main)?

Reply:
Question: if the concern is that after #19806 the active chainstate may change after “cs_main is released in the code immediately above”, then that concern would still exist even if we did not switch from ::ChainstateActive to ActiveChainstate, no? In other words, this is not a new problem that my PR introduced?

As far as I can tell, the behaviour change that switching from ::ChainstateActive to ActiveChainstate introduces is just a possible race condition in ActiveChainstate() whereby a race condition can occur if m_active_chainstate is modified between its assert and its return, which I’d be happy to fix if we think it’s necessary.

Reply:
For what it's worth, I'd figured that the [`GUARDED_BY` on `g_chainman`](https://github.com/bitcoin/bitcoin/blob/master/src/validation.h#L937) would've cascaded through to many of the usages in question, but maybe I misunderstood.

Reply:
>  if the concern is that after #19806 the active chainstate may change after “cs_main is released in the code immediately above”, then that concern would still exist even if we did not switch from ::ChainstateActive to ActiveChainstate, no? In other words, this is not a new problem that my PR introduced?

The concern is that reading/writing `m_active_chainstate` is not guaranteed to be atomic. If this thread reads the variable while another thread is writing to it, then it could read the old value, the new value, or anything else. We have no guarantees at all about what is read if the value is not atomic (or guarded by a mutex to prevent multiple threads accessing it at the same time).

This seems fairly simple to solve - just make `m_active_chainstate` atomic.

Reply:
Fixed by rebasing over #21025, thanks all for the patience and guidance!